app/code/Magento/Customer/Model/PasswordStrengthChecker.php

```diff
@@ -0,0 +1,143 @@
+ * @package Magento\Customer\Model
+ */
+class PasswordStrengthChecker implements PasswordStrengthCheckerInterface
+{
+    /**
+     * Configuration path to customer password minimum length
+     */
+    const XML_PATH_MINIMUM_PASSWORD_LENGTH = 'customer/password/minimum_password_length';
+
+    /**
+     * Configuration path to customer password required character classes number
+     */
+    const XML_PATH_REQUIRED_CHARACTER_CLASSES_NUMBER = 'customer/password/required_character_classes_number';
+
+    /**
+     * @var ResultInterfaceFactory
+     */
+    private $resultFactory;
+
+    /**
+     * @var StringHelper
+     */
+    private $stringHelper;
+
+    /**
+     * @var ScopeConfigInterface
+     */
+    private $scopeConfig;
+
+    /**
+     * PasswordStrengthChecker constructor.
+     * @param StringHelper $stringHelper
+     * @param ScopeConfigInterface $scopeConfig
+     * @param ResultInterfaceFactory $resultFactory
+     */
+    public function __construct(
+        StringHelper $stringHelper,
+        ScopeConfigInterface $scopeConfig,
+        ResultInterfaceFactory $resultFactory
+    ) {
+        $this->stringHelper = $stringHelper;
+        $this->scopeConfig = $scopeConfig;
+        $this->resultFactory = $resultFactory;
+    }
+
+    /**
+     * Check password strength.
+     *
+     * @param string $password
+     * @return ResultInterface
+     */
+    public function check(string $password): ResultInterface
+    {
+        /** @var ResultInterface $result */
+        $result = $this->resultFactory->create(['valid' => true, 'message' => __('')]);
+        $length = $this->stringHelper->strlen($password);
+
+        if ($length > self::MAX_PASSWORD_LENGTH) {
+            $result->setIsValid(false);
+            $result->setMessage(__('Please enter a password with at most %1 characters.', self::MAX_PASSWORD_LENGTH));
+        }
+
+        $minPasswordLength = $this->getMinPasswordLength();
+        if ($length < $minPasswordLength) {
+            $result->setIsValid(false);
+            $result->setMessage(__('Please enter a password with at least %1 characters.', $minPasswordLength));
+        }
+
+        if ($this->stringHelper->strlen(trim($password)) != $length) {
+            $result->setIsValid(false);
+            $result->setMessage(__('The password can\'t begin or end with a space.'));
+        }
+
+        $requiredCharactersCheck = $this->checkRequiredCharacters($password);
+        if ($requiredCharactersCheck !== 0) {
+            $result->setIsValid(false);
+            $result->setMessage(__(
+                'Minimum of different classes of characters in password is %1.' .
+                ' Classes of characters: Lower Case, Upper Case, Digits, Special Characters.',
+                $requiredCharactersCheck
+            ));
+        }
+
+        return $result;
+    }
+
+    /**
+     * Retrieve minimum password length
+     *
+     * @return int
+     */
+    protected function getMinPasswordLength(): int
```

Comment:
Can we mark this method private or remove entirely since it's only used in `check()`?

Reply:
Spot on! I'll make it private.